src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs

```diff
@@ -1004,14 +1112,623 @@ static void Main()
                 //         var a = [1, 2, 3].AsArray();
                 Diagnostic(ErrorCode.ERR_NoSuchMember, "AsArray").WithArguments("collection literals", "AsArray").WithLocation(9, 27));
         }
 
+        [Fact]
+        public void TypeInference_08()
+        {
+            string source = """
+                using System.Collections;
+                using System.Collections.Generic;
+                struct S<T> : IEnumerable<T>
+                {
+                    private List<T> _list;
+                    public void Add(T t)
+                    {
+                        _list ??= new List<T>();
+                        _list.Add(t);
+                    }
+                    public IEnumerator<T> GetEnumerator()
+                    {
+                        _list ??= new List<T>();
+                        return _list.GetEnumerator();
+                    }
+                    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+                }
+                static class Program
+                {
+                    static S<T> AsCollection<T>(this S<T> args)
+                    {
+                        return args;
+                    }
+                    static void Main()
+                    {
+                        var a = AsCollection([1, 2, 3]);
+                        var b = [4].AsCollection();
+                    }
+                }
+                """;
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (27,21): error CS0117: 'collection literals' does not contain a definition for 'AsCollection'
+                //         var b = [4].AsCollection();
+                Diagnostic(ErrorCode.ERR_NoSuchMember, "AsCollection").WithArguments("collection literals", "AsCollection").WithLocation(27, 21));
+        }
+
+        [Fact]
+        public void TypeInference_09()
+        {
+            string source = """
+                using System.Collections;
+                struct S<T> : IEnumerable
+                {
+                    public void Add(T t) { }
+                    IEnumerator IEnumerable.GetEnumerator() => null;
+                }
+                static class Program
+                {
+                    static S<T> AsCollection<T>(this S<T> args)
+                    {
+                        return args;
+                    }
+                    static void Main()
+                    {
+                        _ = AsCollection([1, 2, 3]);
+                        _ = [4].AsCollection();
+                    }
+                }
+                """;
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (15,13): error CS0411: The type arguments for method 'Program.AsCollection<T>(S<T>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
```

Comment:
This error seems misleading? The type arguments absolutely could be inferred, the previous example showed that. The collection literal simply cannot be used to construct `S<T>`.

Reply:
The type arguments cannot be inferred because the *iteration type* of the collection `S<T>` is `object` rather than `T` so no inference is made to `T`.

Reply:
Ah, I see. Thanks for the clarification.